minisat/simp/SimpSolver.cc

```diff
@@ -365,27 +374,28 @@ bool SimpSolver::backwardSubsumptionCheck(bool verbose)
 
         // Search all candidates:
         vec<CRef>& _cs = occurs.lookup(best);
-        CRef*       cs = (CRef*)_cs;
 
-        for (int j = 0; j < _cs.size(); j++)
-            if (c.mark())
+        auto iter = _cs.begin();
+        auto end = _cs.end();
+        while (iter != end) {
```

Comment:
Just thinking... Are we sure that `_cs.end()` is not affected by `removeClause(*iter)` or `strengthenClause(*iter, ~l)`?

Reply:
On a closer look, I am not so sure anymore. I am sure that `removeClause` will not break anything, but `strenghtenClause` is much more complex and can potentially start propagating derived unitary clauses. In case it starts doing that _and_ it ends up propagating to the same var that gave us the `_cs` occurs list, we can assume that the `_cs` would be changed.

I can't say whether this actually can, or cannot happen though.


I also noticed some more low-hanging fruit for improvements in the process.

Reply:
If the while loop was defined as `while (iter != _cs.end()) {` the original semantics would be kept, wouldn't it? Such a guarantee would be fine with me.

Reply:
Not, quite. They would be kept if `_cs` would be made smaller (in other words, no reallocations happen), but not if it is made larger by enough to trigger a realocation. I am going to roll-back this particular change to keep safe.

Reply:
Oh yes, you're right...